lib/shard/src/segment_holder/mod.rs

```diff
@@ -688,7 +697,10 @@ impl SegmentHolder {
                     return Ok(false);
                 }
 
-                let is_applied = if update_nonappendable || write_segment.is_appendable() {
+                let can_apply_operation = !write_segment.is_proxy()
+                    && (update_nonappendable || write_segment.is_appendable());
```

Comment:
I agree that this edge case for proxies is nasty. Let's make segments fully immutable (other than deletes) in a separate PR. ☝️